util/include/channels/mpsc_bounded_channel.h

```diff
@@ -0,0 +1,226 @@
+
+#pragma once
+
+#include <chrono>
+#include <condition_variable>
+#include <memory>
+#include <mutex>
+#include <queue>
+#include <utility>
+
+#include "channel.h"
+
+namespace concord::channel {
+
+// Forward declarations
+template <typename Msg>
+class BoundedMpscSender;
+template <typename Msg>
+class BoundedMpscReceiver;
+
+namespace impl {
+
+// A simple, non-optimized, lock based, bounded MPSC channel.
+//
+// This is an internal type used by the BoundedMpscSender and BoundedMpscReceiver and should not be exposed publicly.
+template <typename Msg>
+class BoundedMpscQueue {
+ public:
+  BoundedMpscQueue(size_t capacity) : capacity_(capacity) {}
+
+ private:
+  template <typename T>
+  friend class ::concord::channel::BoundedMpscSender;
+  template <typename T>
+  friend class ::concord::channel::BoundedMpscReceiver;
+
+  std::optional<size_t> size() {
+    const std::lock_guard<std::mutex> lock(mutex_);
+    return buf_.size();
+  }
+
+  std::optional<size_t> capacity() {
+    // No need for a lock, as capacity_ is immutable
+    return capacity_;
+  }
+
+  void addSender() {
+    const std::lock_guard<std::mutex> lock(mutex_);
+    num_senders_++;
+  }
+
+  void removeSender() {
+    bool num_senders = 0;
+    {
+      const std::lock_guard<std::mutex> lock(mutex_);
+      num_senders_--;
+      num_senders = num_senders_;
+    }
+    if (num_senders == 0) {
+      // Signal the receiver to throw a NoSendersError.
+      cv_.notify_one();
+    }
+  }
+
+  void invalidateReceiver() {
+    const std::lock_guard<std::mutex> lock(mutex_);
+    valid_receiver_ = false;
+  }
+
+  const size_t capacity_;
+  std::queue<Msg> buf_;
+  mutable std::mutex mutex_;
+  std::condition_variable cv_;
+  size_t num_senders_ = 0;
+  bool valid_receiver_ = true;
+};
+
+}  // namespace impl
+
+template <typename Msg>
+class BoundedMpscSender {
+ public:
+  // Try to send a message over a bounded channel.
+  //
+  // Return std::nullopt on sucess.
+  // Return the Msg if the buffer is full
+  //
+  // Throws NoReceiversError if the receiver has been destructed.
+  //
+  // It's important to note that there is a race condition, whereby the receiver can be destructed
+  // immediately after a successful send. There is no absolute guarantee of delivery. The
+  // purpose of returning an error is to prevent endlessly retrying and never noticing that the
+  // channel is effectively destroyed.
+  std::optional<Msg> send(Msg&& msg) {
+    {
+      std::lock_guard<std::mutex> lock(queue_->mutex_);
+      if (!queue_->valid_receiver_) {
+        throw NoReceiversError();
+      }
+      if (queue_->buf_.size() == queue_->capacity_) {
+        return msg;
+      }
+      queue_->buf_.push(std::move(msg));
+    }
+    queue_->cv_.notify_one();
+
+    return std::nullopt;
+  }
+
+  std::optional<size_t> size() { return queue_->size(); }
+  std::optional<size_t> capacity() { return queue_->capacity(); }
+
+  ~BoundedMpscSender() {
+    if (!queue_) return;
```

Comment:
Is that check really necessary? I am asking not because of performance, but just to be sure I don't miss something. I think if a non-null ptr is passed to the ctor (which is a precondition anyway), it will never be null.